### ISSUE TITLE
How to reproduce vanilla NAT performance on raw iwslt16?

### ISSUE THREAD

**Issue by George0828Zhang** [pytorch/fairseq#2187]
#### What is your question?
We can't seem to reproduce the result on non-distilled (original data) IWSLT16 En-De, which is reported as ablation study in [Gu et al. 2018](https://arxiv.org/abs/1711.02281), which has a tokenized BLEU score 16.51.
#### Our results:
|name|best step|SacreBLEU|
|-|-|-|
|teacher(beam1)|8k| 27.81|
|teacher(beam4)|8k| 28.73|
|scratch| 6k|6.33|
|load teacher encoder| 28k|9.71| 

Our teacher is about 1 BLEU below reported (28.89,29.70), but NAT is significantly lower.

#### Notes 
- We use the preprocessed version of IWSLT2016 provided by [Lee et al. 2018](https://arxiv.org/abs/1802.06901) downloaded from [here](https://github.com/nyu-dl/dl4mt-nonauto).
- The above score is reported by fairseq-generate.
- All reported scores are for validation set.
- Models are early stopped after over 10k w/o improvement. 

#### Code
The training args and code to load encoder weights are listed here.
- our model.py
```python
from fairseq.models.nat import NATransformerModel, nonautoregressive_transformer_wmt_en_de
from fairseq.models import register_model, register_model_architecture

import torch

import re
import pdb
import logging

logger = logging.getLogger(__name__)

@register_model("vanilla_gu")
class VanillaGuNAT(NATransformerModel):
    @staticmethod
    def add_args(parser):
        NATransformerModel.add_args(parser)        
        parser.add_argument("--load-encoder-only", action="store_true", #type=bool, nargs='?', const=True, default=False,
        help="whether only load encoder states from checkpoint.")

    def load_state_dict(self, state_dict, strict=True, args=None):
        """Copies parameters and buffers from *state_dict* into this module and
        its descendants.

        Overrides the method in :class:`nn.Module`. Compared with that method
        this additionally "upgrades" *state_dicts* from old checkpoints.
        """
        
        """Overrides fairseq_model.py

        """
        # pdb.set_trace()
        # if getattr(args, "load_encoder_only", False)--gen-subset
        if getattr(args, "load_encoder_only", False):
            logger.warning("Will only load encoder weights!")
            cur = self.state_dict()
            for param in state_dict:
                if re.match(r"^encoder\.", param) is not None:
                    cur[param] = state_dict[param]
            state_dict = cur

        return super().load_state_dict(state_dict, strict=strict, args=args)

@register_model_architecture(
    "vanilla_gu", "vanilla_gu"
)
def vanilla_gu(args):
    nonautoregressive_transformer_wmt_en_de(args)


@register_model_architecture(
    "vanilla_gu", "for_iwslt_16"
)
def for_iwslt_16(args):
    args.encoder_embed_dim = getattr(args, "encoder_embed_dim", 278)
    args.encoder_ffn_embed_dim = getattr(args, "encoder_ffn_embed_dim", 507)
    args.encoder_layers = getattr(args, "encoder_layers", 5)
    args.encoder_attention_heads = getattr(args, "encoder_attention_heads", 2)

    args.decoder_embed_dim = getattr(args, "decoder_embed_dim", 278)
    args.decoder_ffn_embed_dim = getattr(args, "decoder_ffn_embed_dim", 507)
    args.decoder_layers = getattr(args, "decoder_layers", 5)
    args.decoder_attention_heads = getattr(args, "decoder_attention_heads", 2)

    nonautoregressive_transformer_wmt_en_de(args)

```

- Train teacher
```bash
#!/usr/bin/env bash
TASK=teacher
DATA=../DATA/data-bin/iwslt16.lee

mkdir -p checkpoints/$TASK
mkdir -p logdir/$TASK
    
fairseq-train --task translation \
    -s en -t de \
    --max-tokens 4096 \
    --max-tokens-valid 2048 \
    --update-freq 12 \
    --arch transformer \
    --encoder-embed-dim 278 \
    --encoder-ffn-embed-dim 507 \
    --encoder-layers 5 \
    --encoder-attention-heads 2 \
    --decoder-embed-dim 278 \
    --decoder-ffn-embed-dim 507 \
    --decoder-layers 5 \
    --decoder-attention-heads 2 \
    --share-all-embeddings \
    --criterion label_smoothed_cross_entropy \
    --label-smoothing 0.1 \
    --optimizer adam \
    --adam-betas '(0.9, 0.98)' \
    --adam-eps 1e-9 \
    --lr-scheduler inverse_sqrt \
    --warmup-updates 746 \
    --lr 5e-4 \
    --clip-norm 0.0 \
    --dropout 0.1 \
    --weight-decay 0.0001 \
    --no-epoch-checkpoints \
    --eval-bleu \
    --eval-bleu-args '{"beam": 1, "max_len_a": 1.2, "max_len_b": 10}' \
    --eval-bleu-detok moses \
    --eval-bleu-remove-bpe \
    --best-checkpoint-metric bleu --maximize-best-checkpoint-metric \
    --save-dir checkpoints/$TASK \
    --tensorboard-logdir logdir/$TASK \
    --num-workers 4 \
    --max-update 100000 \
    --fp16 \
    --log-format simple --log-interval 1 \
    $DATA
```
- Scratch Model
```bash
#!/usr/bin/env bash
TASK=scratch_vanilla_gu
DATA=../DATA/data-bin/iwslt16.lee

mkdir -p checkpoints/$TASK
mkdir -p logdir/$TASK
    
fairseq-train --task translation_lev_bleu --user-dir . \
    --noise no_noise \
    -s en -t de \
    --max-tokens 4096 \
    --max-tokens-valid 2048 \
    --update-freq 12 \
    --arch for_iwslt_16 \
    --share-all-embeddings \
    --apply-bert-init \
    --src-embedding-copy \
    --pred-length-offset \
    --length-loss-factor 0.1 \
    --criterion nat_loss \
    --label-smoothing 0.0 \
    --optimizer adam \
    --adam-betas '(0.9, 0.98)' \
    --adam-eps 1e-9 \
    --lr-scheduler inverse_sqrt \
    --warmup-updates 746 \
    --lr 5e-4 \
    --clip-norm 0.0 \
    --dropout 0.1 \
    --weight-decay 0.0001 \
    --no-epoch-checkpoints \
    --eval-bleu \
    --eval-bleu-args '{"beam": 1, "max_len_a": 1.2, "max_len_b": 10}' \
    --eval-bleu-detok moses \
    --eval-bleu-remove-bpe \
    --best-checkpoint-metric bleu --maximize-best-checkpoint-metric \
    --save-dir checkpoints/$TASK \
    --tensorboard-logdir logdir/$TASK \
    --num-workers 4 \
    --max-update 100000 \
    --fp16 \
    --log-format simple --log-interval 1 \
    $DATA
```

- Load teacher encoder
```bash
#!/usr/bin/env bash
TASK=load_enc
DATA=../DATA/data-bin/iwslt16.lee
CHECK=../vanillaGu/checkpoints/teacher/checkpoint_best.pt

mkdir -p checkpoints/$TASK
mkdir -p logdir/$TASK
    
fairseq-train --task translation_lev_bleu --user-dir . \
    --noise no_noise \
    -s en -t de \
    --load-encoder-only \
    --restore-file $CHECK \
    --reset-optimizer --reset-dataloader --reset-meters \
    --max-tokens 4096 \
    --max-tokens-valid 2048 \
    --update-freq 12 \
    --arch for_iwslt_16 \
    --share-all-embeddings \
    --apply-bert-init \
    --src-embedding-copy \
    --pred-length-offset \
    --length-loss-factor 0.1 \
    --criterion nat_loss \
    --label-smoothing 0.0 \
    --optimizer adam \
    --adam-betas '(0.9, 0.98)' \
    --adam-eps 1e-9 \
    --lr-scheduler inverse_sqrt \
    --warmup-updates 746 \
    --lr 5e-4 \
    --clip-norm 0.0 \
    --dropout 0.1 \
    --weight-decay 0.0001 \
    --no-epoch-checkpoints \
    --eval-bleu \
    --eval-bleu-args '{"beam": 1, "max_len_a": 1.2, "max_len_b": 10}' \
    --eval-bleu-detok moses \
    --eval-bleu-remove-bpe \
    --best-checkpoint-metric bleu --maximize-best-checkpoint-metric \
    --save-dir checkpoints/$TASK \
    --tensorboard-logdir logdir/$TASK \
    --num-workers 4 \
    --max-update 100000 \
    --fp16 \
    --log-format simple --log-interval 1 \
    $DATA
```

- Evaluate
```bash
DATA=./DATA/data-bin/iwslt16.lee
TASK=scratch_vanilla_gu #load_enc
MODEL=./vanillaGu/checkpoints/${TASK}/checkpoint_best.pt
USER=./vanillaGu/
RESULT=./result/${TASK}/

fairseq-generate \
    --gen-subset valid \
    --task translation_lev \
    --path $MODEL \
    --max-len-a 1.2 \
    --max-len-b 10 \
    --iter-decode-max-iter 1 \
    --iter-decode-eos-penalty 0 \
    --beam 1 \
    --sacrebleu --remove-bpe --tokenizer moses \
    --batch-size 64 \
    --user-dir $USER \
    --results-path $RESULT \
    --model-overrides '{ "load_encoder_only": False }' \
    $DATA
```

#### Note
The task `translation_lev_bleu` is just copied from `translation_lev`, but with few more lines to make --eval-bleu work. 

#### What have you tried?
|tips|BLEU(improvement)|
|-|-|
|Load encoder weights from teacher|9.71 (+3.38)|
|+ Pos-attn mentioned in paper|10.11 (<+1)|
|remove --sacrebleu and --tokenizer|9.68 (no significant difference)| 
|train our own BPE (32k)|no significant difference|

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.5.0
 - OS (e.g., Linux): Ubuntu 20.04 (Linux)
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): pip install ./ (inside cloned fairseq)
 - Python version: 3.8.0
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: RTX2070
 - Any other relevant information: apex installed.

**Comment by da03** [comment]
@George0828Zhang I think iter-decode-max-iter shall be set to 0.

**Comment by George0828Zhang** [comment]
> @George0828Zhang I think iter-decode-max-iter shall be set to 0.

You're right, it should be 0 to not use any refinement steps. How ever i did use `--src-embedding-copy` which rendered `iter-decode-max-iter` irrelevant as in every iteration the input would be the same, copied from source.

**Comment by hzhwcmhf** [comment]
@George0828Zhang Did you solve the problem? I also try to reproduce the vanilla NAT. However, my result on WMT14-ENDE is significantly lower than they reported in "Understanding Knowledge Distillation in Non-autoregressive Machine Translation".

**Comment by speedcell4** [comment]
@hzhwcmhf Did you solve the problem? I am working on this too.

**Comment by hzhwcmhf** [comment]
@speedcell4 I only reproduce WMT14-ENDE. Some keypoints:
* do not use ``--src-embedding-copy``. There are bugs that copy source embedding cannot work with left paddings. See https://github.com/pytorch/fairseq/issues/1316#issuecomment-794108184
* Use large batch size about 64k tokens. You can do it by using multiple GPUs or setting ``update-freq`` > 1.

**Comment by YudiZh** [comment]
> @speedcell4 I only reproduce WMT14-ENDE. Some keypoints:
> 
>     * Do not use `--src-embedding-copy`. There are bugs that copying source embedding cannot work with left paddings. See [ Nonautoregressive-Transformer performs poor #1316 (comment)](https://github.com/pytorch/fairseq/issues/1316#issuecomment-794108184)
> 
>     * Use large batch size about 64k tokens. You can do it by using multiple GPUs or setting `update-freq` > 1.
> 
>     * `iter-decode-max-iter` should be set to 0

Thanks for your tips. I also have another question that what is max-update set with such large batch size to reproduce the experimental results on wmt14.

**Comment by hzhwcmhf** [comment]
@YudiZh 300k updates.   NAT usually needs very long time to converge.

**Comment by YudiZh** [comment]
@hzhwcmhf Thanks a lot for your timely recovery!
